constraint/pkg/client/e2e_test.go

```diff
@@ -0,0 +1,694 @@
+			if _, err := c.AddConstraint(ctx, cstr); err != nil {
+				t.Fatalf("got AddConstraint: %v", err)
+			}
+			obj := &targetData{Name: "Sara", ForConstraint: "Foo"}
+			if _, err := c.AddData(ctx, obj); err != nil {
+				t.Fatalf("got AddData: %v", err)
+			}
+			obj2 := &targetData{Name: "Max", ForConstraint: "Foo"}
+			if _, err := c.AddData(ctx, obj2); err != nil {
+				t.Fatalf("got AddDataX2: %v", err)
+			}
+			rsps, err := c.Audit(ctx, Tracing(true))
+			if err != nil {
+				t.Fatalf("got Audit: %v", err)
+			}
+
+			if len(rsps.ByTarget) == 0 {
+				t.Fatal("got no results by target, want at least one")
+			}
+			for key, r := range rsps.ByTarget {
+				if r.Trace == nil {
+					t.Errorf("got no Trace for: %q, but want trace", key)
+				}
+			}
+		})
+	}
+}
+
+func TestE2EAuditTracingOff(t *testing.T) {
+	for _, tc := range denyAllCases {
+		t.Run(tc.name, func(t *testing.T) {
+			c, err := newTestClient()
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			ctx := context.Background()
+			_, err = c.AddTemplate(ctx, newConstraintTemplate("Foo", tc.rego, tc.libs...))
+			if err != nil {
+				t.Fatalf("got AddTemplate: %v", err)
+			}
+			cstr := newConstraint("Foo", "ph", nil, nil)
+			if _, err := c.AddConstraint(ctx, cstr); err != nil {
+				t.Fatalf("got AddConstraint: %v", err)
+			}
+			obj := &targetData{Name: "Sara", ForConstraint: "Foo"}
+			if _, err := c.AddData(ctx, obj); err != nil {
+				t.Fatalf("got AddData: %v", err)
+			}
+			obj2 := &targetData{Name: "Max", ForConstraint: "Foo"}
+			if _, err := c.AddData(ctx, obj2); err != nil {
+				t.Fatalf("got AddDataX2: %v", err)
+			}
+			rsps, err := c.Audit(ctx, Tracing(false))
+			if err != nil {
+				t.Fatalf("got Audit: %v", err)
+			}
+
+			if len(rsps.ByTarget) == 0 {
+				t.Fatal("got no results by target, want at least one")
+			}
+			for key, r := range rsps.ByTarget {
+				if r.Trace != nil {
+					t.Errorf("got Trace for %q, but want no trace: %v", key, r.Trace)
+				}
+			}
+		})
+	}
+}
+
+func TestE2EDryrunAll(t *testing.T) {
+	for _, tc := range denyAllCases {
+		t.Run(tc.name, func(t *testing.T) {
+			c, err := newTestClient()
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			ctx := context.Background()
+			_, err = c.AddTemplate(ctx, newConstraintTemplate("Foo", `package foo
+violation[{"msg": "DRYRUN", "details": {}}] {
+	"always" == "always"
+}`))
+			if err != nil {
+				t.Fatalf("got AddTemplate: %v", err)
+			}
+			testEnforcementAction := "dryrun"
+			cstr := newConstraint("Foo", "ph", nil, &testEnforcementAction)
+			if _, err := c.AddConstraint(ctx, cstr); err != nil {
+				t.Fatalf("got AddConstraint: %v", err)
+			}
+			rsps, err := c.Review(ctx, targetData{Name: "Sara", ForConstraint: "Foo"})
+			if err != nil {
+				t.Fatalf("got Review: %v", err)
+			}
+
+			got := rsps.Results()
+			want := []*types.Result{{
+				Constraint:        cstr,
+				EnforcementAction: testEnforcementAction,
+				Msg:               "DRYRUN",
```

Comment:
nit: might be nice to have this in the list of constants at the top of the file.  That way, a newcomer (me) would know that's one of the possible states when reading the test file

Reply:
Done!